byterun/win32.c

```diff
@@ -861,3 +861,32 @@ CAMLexport caml_stat_string caml_stat_strdup_of_utf16(const wchar_t *s)
 
   return out;
 }
+
+/* Detect if a named pipe corresponds to MSYS/Cygwin tty: see
+
+   https://github.com/mirror/newlib-cygwin/blob/master/winsup/cygwin/dtable.cc#L932
+*/
+CAMLexport int caml_win32_detect_msys_tty(int fd)
+{
+  char buffer[1024];
+  FILE_NAME_INFO * nameinfo = (FILE_NAME_INFO *) buffer;
+
+  /* check if fd is a pipe */
+  HANDLE h = (HANDLE) _get_osfhandle(fd);
+  if (GetFileType(h) != FILE_TYPE_PIPE)
+    return 0;
+
+  /* get pipe name */
+  if (! GetFileInformationByHandleEx(h, FileNameInfo, buffer, sizeof(buffer) - sizeof(WCHAR)))
+    return 0;
+
+  nameinfo->FileName[nameinfo->FileNameLength / sizeof(WCHAR)] = 0;
```

Comment:
Just for my culture: the cygwin code you link to uses `L'\0'` instead of `0` to terminate widechar strings, are the two equivalent?

Reply:
Yes.